src/hotstuff/protocol.rs

```diff
@@ -5,124 +5,550 @@
 
+    /// Create a new HotStuff instance.
     pub(crate) fn new(
         config: HotStuffConfiguration,
         view_info: ViewInfo,
         sender_handle: SenderHandle<N>,
         validator_set_update_handle: ValidatorSetUpdateHandle<N>,
-        init_validator_set: ValidatorSet,
+        init_validator_set_state: ValidatorSetState,
         event_publisher: Option<Sender<Event>>,
     ) -> Self {
-        let state = HotStuffState::new(
-            config.clone(), 
-            view_info.view, 
-            init_validator_set
-        );
+        let vote_collectors = 
+            <Collectors<VoteCollector>>::new(config.chain_id, view_info.view, &init_validator_set_state);
+        let view_status = ViewStatus::ViewInProgress;
         Self { 
             config,
             view_info,
-            state,
-            sender_handle, 
+            view_status,
+            vote_collectors,
+            sender_handle,
             validator_set_update_handle, 
             event_publisher,
         }
     }
 
+    /// Retrieve the view info currently used by the HotStuff protocol.
+    pub(crate) fn view_info(&self) -> &ViewInfo {
+        &self.view_info
+    }
+
+    /// On receiving new [ViewInfo] from the [Pacemaker](crate::pacemaker::protocol::Pacemaker): send messages
+    /// and perform state updates associated with exiting the current view, update the local view info. 
+    /// 
+    /// This involves the following steps:
+    /// 1. Exit the current view: send a [NewView] message to the leader of the next view.
+    /// 2. Update the internal view info and view status, as well as the vote collectors. 
+    /// 3. If serving as a leader of the newly entered view, propose or nudge.
+    /// 
+    /// ## Precondition
+    /// The [Pacemaker](crate::pacemaker::protocol::Pacemaker) updated the view info.
     pub(crate) fn on_receive_view_info<K: KVStore>(
         &mut self, 
         view_info: ViewInfo,
         block_tree: &mut BlockTree<K>,
         app: &mut impl App<K>,
-    ) {
+    ) -> Result<(), HotStuffError> {
+
+        let validator_set_state = block_tree.validator_set_state()?;
+
+        // 1. Send a NewView message for the current view to the next leader(s).
+        let new_view_msg = 
+            HotStuffMessage::new_view(self.config.chain_id, self.view_info.view, block_tree.highest_qc()?);
+
+        match leaders(self.view_info.view+1, &validator_set_state) {
+            (committed_vs_leader, None) => self.sender_handle.send(committed_vs_leader, new_view_msg.clone()),
+            (committed_vs_leader, Some(prev_vs_leader)) => {
+                self.sender_handle.send(committed_vs_leader, new_view_msg.clone());
+                self.sender_handle.send(prev_vs_leader, new_view_msg.clone())
+            }
+        }
+
+        if let HotStuffMessage::NewView(new_view) = new_view_msg {
+            Event::NewView(NewViewEvent{timestamp: SystemTime::now(), new_view}).publish(&self.event_publisher)
```

Comment:
Nitpick: I prefer something like `event_publisher.publish(event)` over `event.publish(event_publisher)`.

I understand that `event_publisher` right now is an `Option<Sender<Event>>>`, and we can't define methods directly on it. What do you think about wrapping this with a newtype so we can?

Reply:
I am aware that its not possible to define a `publish` method directly on `Option<Sender<Event>>>`, which is why I suggested wrapping the type with a newtype in my original comment.

I personally don't think making this change would be overkill? I acknowledge the benefit to code cleanliness in this case will be small, but I think making the change itself is simple enough that it's worth it (unless making the change is more complex than I envision).

Reply:
Not going to implement.